drivers/mfd/mfd_npm1300.c

```diff
@@ -0,0 +1,85 @@
+/*
+ * Copyright (c) 2023 Nordic Semiconductor ASA
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT nordic_npm1300
+
+#include <errno.h>
+
+#include <zephyr/drivers/i2c.h>
+#include <zephyr/sys/util.h>
+#include <zephyr/drivers/mfd/npm1300.h>
+
+struct mfd_npm1300_config {
+	struct i2c_dt_spec i2c;
+};
+
+static int mfd_npm1300_init(const struct device *dev)
+{
+	const struct mfd_npm1300_config *config = dev->config;
+
+	if (!i2c_is_ready_dt(&config->i2c)) {
+		return -ENODEV;
+	}
+
+	return 0;
+}
+
+int mfd_npm1300_reg_read_burst(const struct device *dev, uint8_t base, uint8_t offset,
+			       void *data, size_t len)
+{
+	const struct mfd_npm1300_config *config = dev->config;
+	uint8_t buff[] = {base, offset};
+
+	return i2c_write_read_dt(&config->i2c, buff, sizeof(buff), data, len);
+}
+
+int mfd_npm1300_reg_read(const struct device *dev, uint8_t base, uint8_t offset, uint8_t *data)
+{
+	return mfd_npm1300_reg_read_burst(dev, base, offset, data, 1U);
+}
+
+int mfd_npm1300_reg_write(const struct device *dev, uint8_t base, uint8_t offset, uint8_t data)
+{
+	const struct mfd_npm1300_config *config = dev->config;
+	uint8_t buff[] = {base, offset, data};
+
+	return i2c_write_dt(&config->i2c, buff, sizeof(buff));
+}
+
+int mfd_npm1300_reg_write2(const struct device *dev, uint8_t base, uint8_t offset, uint8_t data1,
+			   uint8_t data2)
+{
+	const struct mfd_npm1300_config *config = dev->config;
+	uint8_t buff[] = {base, offset, data1, data2};
+
+	return i2c_write_dt(&config->i2c, buff, sizeof(buff));
+}
+
+int mfd_npm1300_reg_update(const struct device *dev, uint8_t base, uint8_t offset, uint8_t data,
+			   uint8_t mask)
+{
+	uint8_t reg;
+	int ret;
+
+	ret = mfd_npm1300_reg_read(dev, base, offset, &reg);
+
+	if (ret < 0) {
+		return ret;
+	}
+
+	reg = (reg & ~mask) | (data & mask);
+
+	return mfd_npm1300_reg_write(dev, base, offset, reg);
+}
```

Comment:
Doesn't this read-modify-write require locking/exclusivity? Or does the architecture of the "sub"-drivers ensure, that no driver (thread) will modify the same register between the read/write calls?

Reply:
I guess it's similar to `i2c_reg_update_byte`, where no bus locking is in place. So all these drivers are for now non-thread safe. The MFD layer could introduce optional bus locking, though.

Reply:
Right, but for `i2c_reg_update_byte()` there is usually just one consumer per I2C device. Here we have multiple consumers.

Reply:
I think multi-threading is going to be quite likely going forward, so I'll add locking to the MFD driver.